lmdeploy/serve/utils.py

```diff
@@ -0,0 +1,253 @@
+            index_range_starts.append(index_range_start)
+            index_range_ends.append(index_range_end)
+
+        def _split_embeddings(input_ids, niter, iter_len, embeddings,
+                              embedding_ranges):
+            embs = [None] * niter
+            ranges = [None] * niter
+
+            if embeddings is None:
+                return embs, ranges
+
+            for i in range(niter):
+                iembs = []
+                iranges = []
+                for emb, (begin, end) in zip(embeddings, embedding_ranges):
+                    assert end <= len(input_ids)
+                    if begin >= (i + 1) * iter_len or end <= i * iter_len:
+                        continue
+                    if isinstance(emb, np.ndarray):
+                        emb = torch.from_numpy(emb)
+                    emb = emb.squeeze()
+                    offx = max(iter_len * i - begin, 0)
+                    offy = max(end - iter_len * (i + 1), 0)
+                    emb = emb[offx:emb.shape[0] - offy]
+                    off = max(begin - iter_len * i, 0)
+                    rng = [off, off + emb.shape[0]]
+                    if self.backend == 'pytorch':
+                        emb = emb.numpy()
+                    iembs.append(emb)
+                    iranges.append(rng)
+
+                iembs = iembs or None
+                iranges = iranges or None
+                embs[i] = iembs
+                ranges[i] = iranges
+
+            return embs, ranges
+
+        if input_embeddings is not None:
+            if not isinstance(input_embeddings[0], list):
+                input_embeddings = [input_embeddings]
+                input_embedding_ranges = [input_embedding_ranges]
+            _input_embeddings = []
+            _input_embedding_ranges = []
+            for i in range(len(input_ids)):
+                embeddings, ranges = _split_embeddings(
+                    input_ids[i], n_max_iter, max_input_len,
+                    input_embeddings[i], input_embedding_ranges[i])
+                _input_embeddings.append(embeddings)
+                _input_embedding_ranges.append(ranges)
+            input_embeddings = _input_embeddings
+            input_embedding_ranges = _input_embedding_ranges
+
+        logits = []
+        generator = self.engine.create_instance()
+        for i in range(n_max_iter):
+            steps = [start[i] for start in index_range_starts]
+            _input_ids = [
+                input_id[start[i]:end[i]] for input_id, start, end in zip(
+                    input_ids, index_range_starts, index_range_ends)
+            ]
+            embeddings = None
+            ranges = None
+            if input_embeddings is not None:
+                embeddings = [x[i] for x in input_embeddings]
+                ranges = [x[i] for x in input_embedding_ranges]
+
+            _logits = generator.decode(_input_ids,
+                                       steps,
```

Comment:
`steps` is optional argument in pytorch engine, show use in kwargs 
https://github.com/InternLM/lmdeploy/blob/fa6f8d2f04fb834a4434e80beac4acc90054ba94/lmdeploy/pytorch/engine/engine_instance.py#L480-L487